astropy_healpix/_core.c

```diff
@@ -0,0 +1,371 @@
+#define NPY_NO_DEPRECATED_API NPY_1_7_API_VERSION
+
+#include <Python.h>
+#include <numpy/arrayobject.h>
+#include <numpy/ufuncobject.h>
+#include "healpix.h"
+#include "healpix-utils.h"
+#include "interpolation.h"
+
+/* FIXME: Remove this once we drop Python 2 support. */
+#include "six.h"
+
+/* FIXME: We need npy_set_floatstatus_invalid(), but unlike most of the Numpy
+ * C API it is only available on some platforms if you explicitly link against
+ * Numpy, which is not typically done for building C extensions. This bundled
+ * header file contains a static definition of _npy_set_floatstatus_invalid().
+ */
+#include "ieee754.h"
+
+
+/* Data structure for storing function pointers for routines that are specific
+ * to the HEALPix ordering scheme. When we create the ufuncs using
+ * PyUFunc_FromFuncAndData, we will set them up to pass a pointer to this
+ * data structure through as the void *data parameter for the loop functions
+ * defined below. */
+typedef struct {
+    int64_t (*order_to_xy)(int64_t, int);
+    int64_t (*xy_to_order)(int64_t, int);
+} order_funcs;
+
+static const order_funcs
+    nested_order_funcs = {healpixl_nested_to_xy, healpixl_xy_to_nested},
+    ring_order_funcs   = {healpixl_ring_to_xy, healpixl_xy_to_ring};
+
+static void *no_ufunc_data[]     = {NULL},
+            *nested_ufunc_data[] = {&nested_order_funcs},
+            *ring_ufunc_data[]   = {&ring_order_funcs};
+
+
+static int64_t nside2npix(int nside)
+{
+    return 12 * ((int64_t) nside) * ((int64_t) nside);
+}
+
+
+static int pixel_nside_valid(int64_t pixel, int nside)
+{
+    return pixel >= 0 && pixel < nside2npix(nside);
+}
+
+
+static void healpix_to_lonlat_loop(
+    char **args, npy_intp *dimensions, npy_intp *steps, void *data)
+{
+    order_funcs *funcs = data;
+    npy_intp i, n = dimensions[0];
+
+    for (i = 0; i < n; i ++)
+    {
+        int64_t pixel = *(int64_t *) &args[0][i * steps[0]];
+        int     nside = *(int *)     &args[1][i * steps[1]];
```

Comment:
Why do we have mostly `int64_t` for integers, but then `int nside`? I guess it doesn't matter because `nside` will fit in an `int`, but wouldn't it be simpler to have `int64_t` everywhere so that we never have to think about what to put as type? - just a thought. Do as you think best and 👍 to merge this in.

Reply:
Such a change would affect the astrometry.net code as well, so that's a bigger question.